beancount/plugins/close_tree.py

```diff
@@ -0,0 +1,98 @@
+want to auto close. An example is the `auto_accounts` plugin that ships with Beancount:
+
+```
+plugin "beancount.plugins.auto_accounts"
+plugin "beancount.plugins.close_tree"
+```
+"""
+
+import time
+from beancount.core import data
+from beancount.core.data import Open, Close
+
+DEBUG = 0
+__plugins__ = ('close_tree',)
+
+
+def close_tree(entries, unused_options_map):
+    """Insert close entries for all subaccounts of a closed account.
+
+    Args:
+      entries: A list of directives. We're interested only in the Open/Close instances.
+      unused_options_map: A parser options dict.
+    Returns:
+      A tuple of entries and errors. """
+
+    start_time = time.time()
+    close_count = 0
+    new_entries = []
+    errors = []
+
+    opens = [e.account for e in entries if isinstance(e, Open)]
+    closes = [e.account for e in entries if isinstance(e, Close)]
+
+    for entry in entries:
+        if isinstance(entry,  Close):
+            subaccounts = [a for a in opens if entry.account + ':' in a and a not in closes]
+            for s in subaccounts:
+                meta = data.new_metadata('<beancount.plugins.close_tree>', 0)
+                close_entry = data.Close(meta, entry.date, s)
+                new_entries.append(close_entry)
+                closes.append(s)
+                close_count += 1
+            if entry.account in opens:
+                new_entries.append(entry)
+        else:
+            new_entries.append(entry)
+
+    # beancount.plugins.auto_accounts, if used in conjunction with this plugin, inserts an Open
+    # entry for empty parent accounts which we are trying to close. For example:
+
+    # plugin "beancount.plugins.auto_accounts"
+    # 2021-06-17 close Expenses:Non-Retirement:Auto:Fit
+    # plugin "beancount.plugins.close_tree"
+    # 2019-01-01 * "Transaction"
+    #   Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
+    #   Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
+    #   Assets:Transfer
+
+    # Above, auto_accounts inserts an Open for 'Expenses:Non-Retirement:Auto:Fit:Gas' on 2021-06-17
+    # since it appears in a close directive. But that's not what we want. The semantic of the close
+    # when u used with this plugin is: "close all children, and close the parent if it was opened
+    # earlier. Below, we filter out entries to correctly handle this scenario:
+
+    opens = [(e.account, e.date) for e in entries if isinstance(e, Open)]
+    closes = [(e.account, e.date) for e in entries if isinstance(e, Close)]
+    retval = []
+    for entry in new_entries:
+        if isinstance(entry, Open) and (entry.account, entry.date) in closes:
+            continue
+        if isinstance(entry, Close) and (entry.account, entry.date) in opens:
+            continue
+        retval.append(entry)
+
+    if DEBUG:
+        elapsed_time = time.time() - start_time
+        print("Close account tree [{:.2f}s]: {} close entries added.".format(elapsed_time, close_count))
```

Comment:
From this I infer that you are worried about the execution time of this plugin. If this is the case, maybe storing the accounts tree in a flat list is not the best choice. You could use a tree data structure. Also iterating the list of entries six times is not optimal. If there are no performance issues, I would delete the debug code, as it is anyhow not discoverable or accessible by the users and there are better profiling tools.